src/main.ts

```diff
@@ -0,0 +1,159 @@
+import * as cp from 'child_process'
+import * as os from 'os'
+import { AutoLanguageClient, ActiveServer } from 'atom-languageclient'
+import * as path from 'path'
+import { CompositeDisposable, FilesystemChangeEvent } from 'atom';
+import * as languageServer from 'ocaml-language-server'
+import * as fs from 'fs-extra'
+import merge from 'deepmerge'
+import throttle from 'lodash.throttle'
+import * as pkg from '../package.json'
+import config from './config'
+import * as Utils from './utils'
+import { DeepPartial, Config } from './types'
+
+const confFile = ".atom/ide-reason.json"
+const defaultConfig = languageServer.ISettings.defaults.reason
+
+const scopes = [
+  'source.reason',
+  'source.ocaml',
+  'source.re',
+  'source.ml',
+  'source.rei',
+  'source.mli',
+]
+
+class ReasonMLLanguageClient extends AutoLanguageClient {
+  subscriptions: CompositeDisposable | null = null
+  confPerProject: DeepPartial<Config> = {}
+  servers: { [K: string]: ActiveServer } = {}
+
+  getGrammarScopes () { return scopes }
+  getLanguageName () { return 'Reason' }
+  getServerName () { return 'ocamlmerlin' }
+  getConnectionType() { return 'ipc' as 'ipc' }
+  getRootConfigurationKey() { return 'ide-reason' }
+
+  mapConfigurationObject(config: Config) {
+    config = merge(merge({}, config), this.confPerProject)
+
+    return {
+      reason: {
+        codelens: {
+          enabled: false,
+          unicode: true,
+        },
+        debounce: config.debounce,
+        diagnostics: {
+          tools: config.diagnostics.tools,
+          merlinPerfLogging: false,
+        },
+        format: config.format,
+        path: config.path,
+        server: config.server,
+      },
+    }
+  }
+
+  activate() {
+    super.activate()
+    require('atom-package-deps').install('ide-reason')
+      .then(() => {
+        console.log('All dependencies installed, good to go')
+      })
+
+    this.subscriptions = new CompositeDisposable();
+
+    this.subscriptions.add(
+      atom.commands.add('atom-workspace', {
+        [`${pkg.name}:generate-config`]: () => this.generateConfig().catch(console.error)
+      }),
+      atom.commands.add('atom-workspace', {
+        [`${pkg.name}:restart-all-servers`]: () => this.restartAllServers().catch(console.error)
+      }),
+    )
+
+    const handleConfFileChange = throttle((events: FilesystemChangeEvent) => {
```

Comment:
BTW why do you throttle here? AFAIU `atom.project.onDidChangeFiles` notifies us on change (when user saves file) so `handleConfFileChange` is one time operation we can fire right away?

Reply:
You are right, it's not necessary...